File: Linear_Control_Course/Chapter19/Lesson4/Chapter19_Lesson4.m
% Plant: second-order joint model G(s) = Km / (J s^2 + B s)
J = 0.01;
B = 0.1;
Km = 0.5;

s = tf('s');
G = Km / (J * s^2 + B * s);

% Lead design (parameters from Bode-based procedure)
wc_des = 10;      % rad/s
alpha = 0.2;      % z1/p1 = alpha < 1

z1 = wc_des / sqrt(alpha);
p1 = wc_des * sqrt(alpha);

G_lead = (s + z1) / (s + p1);

% Choose gain Kc so that |L(j wc_des)| ≈ 1
[magG, ~] = bode(G, wc_des);
[magLead, ~] = bode(G_lead, wc_des);
magG = squeeze(magG);
magLead = squeeze(magLead);
Kc = 1 / (magG * magLead);

% Lag design for improved low-frequency error
beta = 10;
wz2 = wc_des / 10;
wp2 = wz2 / beta;
G_lag = (s + wz2) / (s + wp2);

Gc = Kc * G_lead * G_lag;
L = Gc * G;
T = feedback(L, 1);

figure;
margin(L);
title('Bode and margins of lead-lag compensated open-loop');

figure;
step(T);
grid on;
title('Closed-loop step response with lead-lag compensator');
xlabel('Time [s]');
ylabel('Joint position [rad]');

% In Simulink:
% - Implement G(s) as a Transfer Fcn block.
% - Implement G_lead(s) and G_lag(s) as cascaded Transfer Fcn blocks.
% - Use a Step block for reference and a Sum block for the error.
% - Optionally interface with Robotics System Toolbox rigidBodyTree models.

File: Tutorials/Codes/Linear_Control_Course/Chapter19/Lesson4/Chapter19_Lesson4.py
import numpy as np
import control as ctl
# If needed: pip install control
# For robotics context: pip install roboticstoolbox-python

# --- Plant: 1-DOF joint approximation ---
J = 0.01   # kg m^2
B = 0.1    # N m s/rad
Km = 0.5   # N m/A (effective gain)

s = ctl.TransferFunction.s
G = Km / (J * s**2 + B * s)

# --- Lead design (numbers chosen to give extra phase near wc) ---
wc_des = 10.0   # desired crossover rad/s
alpha = 0.2     # z1/p1 = alpha < 1 for lead

z1 = wc_des / np.sqrt(alpha)
p1 = wc_des * np.sqrt(alpha)

G_lead = (s + z1) / (s + p1)

# Preliminary gain so that |Kc * G_lead(j wc) * G(j wc)| ≈ 1
# Compute magnitude at wc_des
mag_plant, phase_plant, w = ctl.bode(G, [wc_des], Plot=False)
mag_lead, phase_lead, _ = ctl.bode(G_lead, [wc_des], Plot=False)
Kc = 1.0 / (mag_plant[0] * mag_lead[0])

# --- Lag design to improve steady-state position error ---
# Suppose we want 10x improvement in low-frequency error constants
beta = 10.0
wz2 = wc_des / 10.0
wp2 = wz2 / beta

G_lag = (s + wz2) / (s + wp2)

# Full controller: lead-lag with gain Kc
Gc = Kc * G_lead * G_lag

# Open-loop and closed-loop
L = Gc * G
T = ctl.feedback(L, 1)  # unity feedback

# Bode plot of compensated open-loop
ctl.bode(L, dB=True)

# Step response of joint position
t = np.linspace(0, 2.0, 1000)
t, y = ctl.step_response(T, T=t)

import matplotlib.pyplot as plt
plt.figure()
plt.plot(t, y)
plt.xlabel("Time [s]")
plt.ylabel("Joint position (rad)")
plt.title("Closed-loop step response with lead-lag compensator")
plt.grid(True)
plt.show()
